### ISSUE TITLE
Handle errors without using rescue/catch

### ISSUE THREAD

**Issue by CrowdHailer** [OleMchls/elixir-hpack#10]
Is there any interface that exists, (or would be trivial to implement) so that errors could be handled via `:ok/:error` tuples? Ideally one that has the context/table just as an object in the same process.

ie.
```
{:ok, {headers, table}} = HPack.decode(header_block, table)
{:error, :malformed_block} = HPack.decode(header_block, table)
```

**Comment by nesQuick** [comment]
I wonder if and how this could fir into the "let it crash" theory, because as of now this lib only covers the happy path.

**Comment by CrowdHailer** [comment]
Bad input is not something that should trigger a let it crash result. I am testing my server against the h2spec and it sends some invalid header_blocks and expects a go_away frame to be sent. the only way this can work at the moment for me is to use a try catch block from which i return :ok/:error

**Comment by nesQuick** [comment]
The only solution I can think of is to have this catch within the `decode` or `parse` function. Or checking the input (which I would dislike on the first grasp). Would you have another idea?

**Comment by CrowdHailer** [comment]
You should not need a catch. part of the parsing should be able to have a final decode function head that just returns an error and where possible a reason, possible just describing the shape it should have had to match one of the previous clauses

**Comment by nesQuick** [comment]
I think the API you outlined in your initial post looks good, I'd love to see your take on it :)

**Comment by CrowdHailer** [comment]
Hey, I've moved my project to :cowlib for the moment, all in a nice wrapper so its easy to change.
Pretty busy at the moment so that will probably stay but I would like to move back to an all elixir implementation so it could be a nice project for beginners to look at

**Comment by OleMchls** [comment]
Solved with #13 thanks @lucacorti
